### PR TITLE
Kamelets

### LINKED ISSUE #375: kamelets: create a camel-kamelet component
As described in https://github.com/apache/camel-k/issues/1567, a _kamelet_ should be consumed like a regular component, like

```java
from("timer:trigger")
    .to("kamelet:myKamelet?foo=bar")
    .log("${body}")
```

### Mechanic
The `createEndpoint` of the `kamelet` component should looks like:

```java
Endpoint createEndpoint(String uri, String remaining, Map<String, Object> parameters) {
    var kameletName = StringHelper.before(reamining, "/");
    var kameletId = StringHelper.after(reamining, "/");
    if (kameletId == null) {
        kameletId = generateKameletId();
    }

    // load properties
    var options = new HashMap<String, Object>()
    options.putAll(getProperties("camel.kamelet." + kameletName))
    options.putAll(getProperties("camel.kamelet." + kameletName + "." + kameletId))
    options.putAll(parameters)

    // materialize a route builder by loading the kamelet as a standard source, so through
    // one of the available org.apache.camel.k.SourceLoader      
    var route = loadRoute(kameletName)
 
    // replace the input with a direct
    route.setInput("direct:" + kameletName + "-" + kameletId)

    // set route specific options
    // requires https://issues.apache.org/jira/browse/CAMEL-14963
    route.setPropertiers(options)

    // create an endpoint that connects the route create above
    return context.getEndpoint("direct:" + kameletName + "-" + kameletId)
}
```

### PR DESCRIPTION
**Pull request by lburgazzoli**
Fixes #375 
Fixes https://github.com/apache/camel-k/issues/1575

**Release Note**
```release-note
Add support for kamelets 
- https://github.com/apache/camel-k-runtime/issues/375
- https://github.com/apache/camel-k/issues/1567
```

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

File: camel-k-runtime-core/src/main/java/org/apache/camel/k/support/PropertiesSupport.java [hunk 1]
```diff
@@ -26,34 +26,84 @@
...
+    public static <T> T bindProperties(CamelContext context, T target, Predicate<String> filter, String prefix, boolean stripPrefix) {
         final PropertiesComponent component = context.getPropertiesComponent();
-        final Properties properties = component.loadProperties(k -> k.startsWith(prefix));
+        final Properties propertiesWithPrefix = component.loadProperties(filter);
+        final Map<String, Object> properties = new HashMap<>();
+
+        propertiesWithPrefix.stringPropertyNames().forEach(
+            name -> properties.put(
+                // PropertyBindingSupport does not support dash properties
+                StringHelper.dashToCamelCase(stripPrefix ? name.substring(prefix.length()) : name),
```
Comment (davsclaus) [inline camel-k-runtime-core/src/main/java/org/apache/camel/k/support/PropertiesSupport.java:72]: Yes it ought to support this

File: camel-k-runtime-core/src/generated/resources/META-INF/services/org/apache/camel/configurer/SourceDefinition [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Generated by camel build tools - do NOT edit this file!
+class=org.apache.camel.k.SourceDefinitionConfigurer
```
Comment (davsclaus) [inline camel-k-runtime-core/src/generated/resources/META-INF/services/org/apache/camel/configurer/SourceDefinition:2]: camel now generate these via FQN name, so just beware of deleting this old and adding the new

File: camel-k-runtime-core/src/main/java/org/apache/camel/k/Sources.java [hunk 3]
```diff
@@ -18,177 +18,119 @@
...
 import java.util.Base64;
 import java.util.Collections;
 import java.util.List;
-import java.util.Map;
 import java.util.Optional;
 import java.util.UUID;
 import java.util.zip.GZIPInputStream;
 
 import org.apache.camel.CamelContext;
 import org.apache.camel.k.support.StringSupport;
```
Comment (davsclaus) [inline camel-k-runtime-core/src/main/java/org/apache/camel/k/Sources.java:29]: I wonder if some of the StringSupport you have here can use what we have in camel-core, and if not maybe move those to camel-util etc.

File: camel-k-runtime-core/src/main/java/org/apache/camel/k/support/PropertiesSupport.java [hunk 4]
```diff
@@ -26,34 +26,84 @@
...
+
+        PropertyConfigurer configurer = null;
+        if (target instanceof Component) {
+            // the component needs to be initialized to have the configurer ready
+            ServiceHelper.initService(target);
+            configurer = ((Component) target).getComponentPropertyConfigurer();
+        }
+
+        if (configurer == null) {
+            String name = target.getClass().getSimpleName();
```
Comment (davsclaus) [inline camel-k-runtime-core/src/main/java/org/apache/camel/k/support/PropertiesSupport.java:84]: Using long name now, see the code in camel-core

File: camel-kamelet/src/generated/resources/org/apache/camel/component/kamelet/kamelet.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
+{
+  "component": {
+    "kind": "component",
+    "name": "kamelet",
+    "title": "Kamelet",
+    "description": "The Apache Software Foundation provides support for the Apache community of open-source software projects.\n    The Apache projects are characterized by a collaborative, consensus based development process, an open and\n    pragmatic software license, and a desire to create high quality software that leads the way in its field.\n    We consider ourselves not simply a group of projects sharing a server, but rather a community of developers\n    and users.",
```
Comment (davsclaus) [inline camel-kamelet/src/generated/resources/org/apache/camel/component/kamelet/kamelet.json:6]: The description here is wrong,

File: camel-kamelet/src/main/java/org/apache/camel/component/kamelet/KameletEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            if (producer != null) {
+                return producer.process(exchange, callback);
+            } else {
+                callback.done(true);
+                return true;
+            }
+        }
+
+        @Override
+        protected void doStart() throws Exception {
```
Comment (davsclaus) [inline camel-kamelet/src/main/java/org/apache/camel/component/kamelet/KameletEndpoint.java:160]: You can maybe get the endpoint and create producer in doInit and the same for the consumer

File: camel-kamelet/src/main/java/org/apache/camel/component/kamelet/KameletEndpoint.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+import org.apache.camel.spi.Metadata;
+import org.apache.camel.spi.UriEndpoint;
+import org.apache.camel.spi.UriPath;
+import org.apache.camel.support.DefaultAsyncProducer;
+import org.apache.camel.support.DefaultConsumer;
+import org.apache.camel.support.DefaultEndpoint;
+import org.apache.camel.support.service.ServiceHelper;
+import org.apache.camel.util.ObjectHelper;
+
+@UriEndpoint(
```
Comment (davsclaus) [inline camel-kamelet/src/main/java/org/apache/camel/component/kamelet/KameletEndpoint.java:38]: Should be lenientProperties=true as additional parameters are used as template parameters, arent they?

File: camel-kamelet/src/test/resources/log4j2-test.xml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+<Configuration status="INFO">
+  <Appenders>
+    <Console name="STDOUT" target="SYSTEM_OUT">
+      <PatternLayout pattern="%d{yyyy-MM-dd HH:mm:ss.SSS}|%-5level|%t|%c{1} - %msg%n"/>
+    </Console>
+    <Null name="NONE"/>
+  </Appenders>
+
+  <Loggers>
+    <Logger name="org.apache.camel.component.kamelet" level="TRACE"/>
```
Comment (davsclaus) [inline camel-kamelet/src/test/resources/log4j2-test.xml:29]: Would be nice to log to a file by default

File: camel-k-runtime-core/src/generated/resources/META-INF/services/org/apache/camel/configurer/SourceDefinition (hunk 2 above)
Comment (lburgazzoli) [reply camel-k-runtime-core/src/generated/resources/META-INF/services/org/apache/camel/configurer/SourceDefinition:2]: I'm on it

File: camel-k-runtime-core/src/main/java/org/apache/camel/k/Sources.java (hunk 3 above)
Comment (lburgazzoli) [reply camel-k-runtime-core/src/main/java/org/apache/camel/k/Sources.java:29]: yeah that's something we have been discussing yesterday, my plan is to move whatever generic I have here to camel

File: camel-k-runtime-core/src/main/java/org/apache/camel/k/support/PropertiesSupport.java (hunk 4 above)
Comment (lburgazzoli) [reply camel-k-runtime-core/src/main/java/org/apache/camel/k/support/PropertiesSupport.java:84]: on it

File: camel-kamelet/src/generated/resources/org/apache/camel/component/kamelet/kamelet.json (hunk 5 above)
Comment (lburgazzoli) [reply camel-kamelet/src/generated/resources/org/apache/camel/component/kamelet/kamelet.json:6]: where it comes from ?

Comment (davsclaus) [reply camel-kamelet/src/generated/resources/org/apache/camel/component/kamelet/kamelet.json:6]: It takes the javadoc from the endpoint class, and fallback to pom.xml description

Comment (lburgazzoli) [reply camel-kamelet/src/generated/resources/org/apache/camel/component/kamelet/kamelet.json:6]: ah ok, so here it takes the description from the apache parent pom :) I'll fix it

File: camel-kamelet/src/main/java/org/apache/camel/component/kamelet/KameletEndpoint.java (hunk 6 above)
Comment (lburgazzoli) [reply camel-kamelet/src/main/java/org/apache/camel/component/kamelet/KameletEndpoint.java:160]: I can probably do it for the producer but the consumer needs the processor so I can't eagerly create it right ?

**Pull request merged**